alibi_detect/utils/pytorch/kernels.py

```diff
@@ -1,8 +1,44 @@
+from abc import abstractmethod
 import numpy as np
 import torch
 from torch import nn
 from . import distance
-from typing import Optional, Union, Callable
+from typing import Optional, Union, Callable, List
+from copy import deepcopy
+
+
+def infer_kernel_parameter(
+    kernel: 'BaseKernel',
+    x: torch.Tensor,
+    y: torch.Tensor,
+    dist: torch.Tensor,
+    infer_parameter: bool = True
+) -> None:
+    """
+    Infer the kernel parameter from the data.
+
+    Parameters
+    ----------
+    kernel
+        The kernel function.
+    x
+        Tensor of instances with dimension [Nx, features].
+    y
+        Tensor of instances with dimension [Ny, features].
+    dist
+        Tensor with dimensions [Nx, Ny], containing the pairwise distances between `x` and `y`.
+    infer_parameter
+        Whether to infer the kernel parameter.
+    """
+    if kernel.trainable and infer_parameter:
```

Comment:
This relates back to @ojcobb 's comment on the possible requirement to set `trainable` on a parameter by parameter basis. It is possible that we do not allow for this just yet in this PR, but then a GitHub issue should be opened to follow up.

Reply:
Currently, we allow each parameter to be configured separately on `trainable` for a torch and `requires_grad` for tf.  This `kernel.trainable` was used to provide an accessble attribute for other methods to determine the training requirements. Guess it would be better here that I just do a quick implementation that checks all the parameters and returns if at least one parameter requires training.